src/substrait/type_inference.py

```diff
@@ -0,0 +1,202 @@
+        return stt.Type(
+            bool=stt.Type.Boolean(nullability=stt.Type.Nullability.NULLABILITY_NULLABLE)
+        )
+    elif rex_type == "nested":
+        return infer_nested_type(expression.nested)
+    elif rex_type == "subquery":
+        subquery_type = expression.subquery.WhichOneof('subquery_type')
+
+        if subquery_type == "scalar":
+            scalar_rel = infer_rel_schema(expression.subquery.scalar.input)
+            return scalar_rel.types[0]
+        elif subquery_type == "in_predicate" or subquery_type == "set_comparison" or subquery_type == "set_predicate":
+            stt.Type.Boolean(nullability=stt.Type.Nullability.NULLABILITY_NULLABLE) # can this be a null?
+        else:
+            raise Exception(f"Unknown subquery_type {subquery_type}")
+    else:
+        raise Exception(f"Unknown rex_type {rex_type}")
+
+def infer_rel_schema(rel: stalg.Rel) -> stt.Type.Struct:
+    rel_type = rel.WhichOneof("rel_type")
+
+    if rel_type == "read":
+        (common, struct) = (rel.read.common, rel.read.base_schema.struct)
+    elif rel_type == "filter":
+        (common, struct) = (rel.filter.common, infer_rel_schema(rel.filter.input))
+    elif rel_type == "fetch":
+        (common, struct) = (rel.fetch.common, infer_rel_schema(rel.fetch.input))
+    elif rel_type == "aggregate":
+        parent_schema = infer_rel_schema(rel.aggregate.input)
+        grouping_types = [infer_expression_type(g, parent_schema) for g in rel.aggregate.grouping_expressions]
+        measure_types = [m.measure.output_type for m in rel.aggregate.measures]
+
+        raw_schema = stt.Type.Struct(
+            types=grouping_types + measure_types,
+            nullability=parent_schema.nullability,
+        )
+
+        (common, struct) = (rel.aggregate.common, raw_schema)
+    elif rel_type == "sort":
+        (common, struct) = (rel.sort.common, infer_rel_schema(rel.sort.input))
+    elif rel_type == "project":
+        parent_schema = infer_rel_schema(rel.project.input)
+        expression_types = [
+            infer_expression_type(e, parent_schema) for e in rel.project.expressions
+        ]
+        raw_schema = stt.Type.Struct(
+            types=list(parent_schema.types) + expression_types,
+            nullability=parent_schema.nullability,
+        )
+
+        (common, struct) = (rel.project.common, raw_schema)
+    elif rel_type == "set":
+        (common, struct) = (rel.fetch.common, infer_rel_schema(rel.set.inputs[0]))
+    elif rel_type == "cross":
+        left_schema = infer_rel_schema(rel.cross.left)
+        right_schema = infer_rel_schema(rel.cross.right)
+
+        raw_schema = stt.Type.Struct(
+            types=left_schema.types + right_schema.types,
+            nullability=stt.Type.Nullability.NULLABILITY_REQUIRED,
+        )
+
+        (common, struct) = (rel.cross.common, raw_schema)
+    else:
+        raise Exception(f"Unhandled rel_type {rel_type}")
```

Comment:
This includes join which are complicated as it outputs different number of fields based on the join type.  Fine to skip that for now.

Reply:
I added join cases as well while we're at it 😄 Direct Output Order for Join Operation should probably be changed in the docs to reflect these cases, it's a bit fuzzy now.